quinn-proto/src/connection/assembler.rs

```diff
@@ -98,34 +108,106 @@ impl Assembler {
 
         let new = BinaryHeap::with_capacity(self.data.len());
         let old = mem::replace(&mut self.data, new);
+        self.buffered = fragmented_buffered;
         for chunk in old.into_sorted_vec().into_iter().rev() {
-            let end = offset + (buffer.len() as u64);
-            if let Some(overlap) = end.checked_sub(chunk.offset) {
-                if let Some(bytes) = chunk.bytes.get(overlap as usize..) {
-                    buffer.extend_from_slice(bytes);
-                }
-            } else {
-                let bytes = buffer.split().freeze();
-                self.data.push(Buffer { offset, bytes });
+            if chunk.is_defragmented() {
+                self.buffered += chunk.size;
+                self.data.push(chunk);
+                continue;
+            } else if !chunk.should_defragment() {
+                self.buffered += chunk.bytes.len();
```

Comment:
Currently, it is not fully identical (because `chunk.size` could differ from `chunk.bytes.len()`), but I will get rid of the `size` field and turn this into one `if` condition.

Reply:
Fixed.